### ISSUE TITLE
[class.spaceship] lacks to specify the behavior if the defined expression for synthesized three-way comparison is ill-formed

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/draft#5338]
Consider this example:
````cpp
struct C{
   int operator <=>(C const&) const noexcept = default;
   int a = 0;
};
int main(){
    using type = decltype(&C::operator<=>);
}
````

GCC:  
````
<source>:8:39: error: use of deleted function 'constexpr int C::operator<=>(const C&) const'
    8 |     using type = decltype(&C::operator<=>);
      |                                       ^~~
<source>:4:8: note: 'constexpr int C::operator<=>(const C&) const' is implicitly deleted because the default definition would be ill-formed:
    4 |    int operator <=>(C const&) const noexcept = default;
      |        ^~~~~~~~
<source>:5:8: error: three-way comparison of 'C::a' has type 'std::strong_ordering', which does not convert to 'int'
    5 |    int a = 0;
````
GCC and Clang [have](https://godbolt.org/z/jcv4bGoGG) a divergence implementation on the point that if the defined expression is ill-formed. Specifically, the expanded list of subobjects for an object x of type C consists of {C::a}, where the synthesized three-way comparison is defined as `static_cast<int>(a<=>a)` as per [class.spaceship] p1.1. Since the operand of the operator `<=>` in the defined expression has type `int`, built-in expression interpretation applies to the operator, its type is std​::​strong_­ordering whose prvalue cannot be converted to type `int`. Thus, the defined expression `static_cast<int>(a<=>a)` is ill-formed, on this point, GCC thinks the defaulted three-way comparison function should be deleted while Clang thinks it's ok. However, [class.spaceship.note] p1 wants to imply something regarding this case. However, it's not a formal rule and no overload resolution performs in this case too.

**Comment by frederick-vs-ja** [comment]
Isn't this the case for `C<Weak>` in [P1186R3](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1186r3.html#what-does-it-mean-to-require-an-operation)?

**Comment by xmh0511** [comment]
> Isn't this the case for `C<Weak>` in [P1186R3](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1186r3.html#what-does-it-mean-to-require-an-operation)?

Yes, that paper lists large particular cases that can expose the problem to which this issue is concerning. From [class.spaceship.note] p1, the standard prefers to consider these synthesized three-way comparisons to be ill-formed, thus I agree with the opinion in that paper: Erring on the side of **ill-formed** helps programmers catch bugs earlier; delete the defaulted function instead.

**Comment by xmh0511** [comment]
The proposal in that paper says:
> If the declared return type of a defaulted three-way comparison operator function is R and the synthesized three-way comparison for comparison category type R between any objects xi and xi is not defined or **would be ill-formed**, the operator function is defined as deleted.

Conversely, the current draft only talks about "is not defined". For the example given in this issue, it has a defined synthesized three-way comparison, merely the defined expression is ill-formed. The case falls into the "would be ill-formed", however, the current draft lacks to specify such a condition in the formal rules.

**Comment by jensmaurer** [comment]
Not an improvement. "If ill-formed, then deleted" is too far-reaching to be implementable.

See CWG2432.

**Comment by frederick-vs-ja** [comment]
Using "would be ill-formed" here is not an improvement but a regression.

The "would be ill-formed" part was removed by [P2002R1](https://wg21.link/p2002r1), following the rationale described in [P1630R1](https://wg21,link/p1630r1): no function should be instantiated when determining whether a defaulted function is deleted.

**Comment by xmh0511** [comment]
@jensmaurer @frederick-vs-ja 

The other opinion is, whether the defaulted function is defined as deleted, which is only determined by whether the synthesized three-way comparison is defined, keeps the same as the current draft did. However, if once the defaulted function is odr-used or its definition is required, the prior synthesized three-way comparison is rechecked to determine whether the program is well-formed as if the synthesized three-way comparison would appear in the definition of the defaulted function.

This is consistent with what Clang has done.

**Comment by jensmaurer** [comment]
And that's consistent with what we do for defaulted constructors: In a first phase, we check whether the default constructor is deleted; in a second phase (when the definition is needed), we compile the actual definition (which might end up being ill-formed).

And this seems to be the intent here, too, and seems to be supported by the existing wording, so maybe it's enough to add a note.

**Comment by xmh0511** [comment]
@jensmaurer For default constructor or other special member functions, we just explicitly define the condition that the defaulted default constructor should be defined as deleted. For ill-formed cases, I think they are regulated by [class.base.init] together with [dcl.init]. Such as the default initialization of the member of reference type. we **explicitly** state such a case will be ill-formed through [dcl.init.general] p10.   

The difference here is that we didn't explicitly phrase what will happen when evaluating the defaulted three-way comparison except that we may imply it through specifying the return value of the comparison in [class.spaceship] p3. Maybe, we could explicitly state the "ill-formed" story in [class.spaceship] p3:
>until the first index i where the synthesized three-way comparison of type R between x<sub>i</sub> and y<sub>i</sub> yields a result value v<sub>i</sub> where v<sub>i</sub> != 0, contextually converted to bool, yields true; V is a copy of v<sub>i</sub>.

It seems that the rule intends to say the evaluation of the default comparison will evaluate the synthesized comparison of the subobjects. However, it's not explicit. Maybe, we could say: if any  synthesized three-way comparison of type R between x<sub>i</sub> and y<sub>i</sub> results a valid expression, the program is ill-formed. I think this sentence is close to what the implementation did.

**Comment by jensmaurer** [comment]
The point is still that the constructor descriptions rely on specifications elsewhere to get to the ill-formed cases. I'm not seeing why we need to be normatively different for the comparison operators. I do agree that the "deleted" and "use this expression" cases are a bit more interwoven for the comparison functions.

**Comment by xmh0511** [comment]
@jensmaurer  I think we just need to specify what expressions are in the implicit definition of a defaulted comparison operator function, all the issues mentioned above will be resolved. 

Consider [except.spec] p11
>The exception specification for a comparison operator function ([over.binary]) without a noexcept-specifier that is defaulted on its first declaration is potentially-throwing if and only if **any expression in the implicit definition** is potentially-throwing.

we seem to haven't defined the concept for expression in the implicit definition. For a defaulted three-way comparison, the expressions presumably are these synthesized three-way comparisons for the subobjects in the object x of class C of which the defaulted function is a member. The same is true of defaulted equality operator function. For these defaulted functions, we just specify their return values. We never explicitly phrase that these expressions that are used to determine the return value appear within the implicit definitions.


So, I think we lack to specify the concept for what expressions are in the implicit definition of a defaulted function.
